---
title: "DESeq2 analysis of samples two vs two"
output: html_notebook
---

This notebook tries to explain the protocol to compare samples two vs two in order to determine which ASVs/Taxons are significatively abundant in one of the samples. As we are comparing abundance among two samples we can use the same library that compare gene counts for RNASeq, that is, DESeq2. From now on I am going to write down all the commands that must be run in R though first of all you must specify the working directory with setwd command.

Let's start loading all the libraries we are going to use to convert QIIME2 artifacts files (qiime2R), process them as phylogenetic data (phyloseq), manipulate tables (readr, tibble, dplyr and tidyr), and to plot the results (ggplot2)

```{r libraries}
library("phyloseq")
library("qiime2R")
library("DESeq2")
library("readr")
library("tibble")
library("dplyr")
library("tidyr")
library("ggplot2")
```


First of all we must load the artifact files we have obtained in QIIME2 